feat: add autoware_rosbag2_anonymizer usage

StepTurtle opened the pull request:
## Description

This PR updates the dataset documentation and adds instructions for the data anonymizer tool.

Related issue:
- https://github.com/autowarefoundation/autoware/issues/4551

Repo Link:
- https://github.com/autowarefoundation/autoware_rosbag2_anonymizer

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [X] I've confirmed the [contribution guidelines].
- [X] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The Reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

mitsudome-r commented on docs/datasets/data-anonymization/index.md:
```diff
@@ -0,0 +1,221 @@
+
+## Overview
+
+Autoware provides a tool to anonymize ROS2 bag files.
```
```suggestion
Autoware provides a tool to anonymize ROS 2 bag files.
```

mitsudome-r commented on docs/datasets/data-anonymization/index.md:
```diff
@@ -0,0 +1,221 @@
+
+!!! warning
+
+    Be sure you have installed the ROS2 on your system.
```
```suggestion
    Be sure you have installed the ROS 2 on your system.
```

mitsudome-r commented on docs/datasets/data-anonymization/index.md:
```diff
@@ -0,0 +1,221 @@
+
+## Usage
+
+The tool provides two options to anonymize images in ROS2 bag files.
```
```suggestion
The tool provides two options to anonymize images in ROS 2 bag files.
```

mitsudome-r commented on docs/datasets/data-anonymization/index.md:
```diff
@@ -0,0 +1,221 @@
+
+#### Step 1: Create a Dataset
+
+Create an initial dataset with the unified model. You can provide multiple ROS2 bag files to create a dataset. After
```
```suggestion
Create an initial dataset with the unified model. You can provide multiple ROS 2 bag files to create a dataset. After
```

mitsudome-r commented on docs/datasets/data-anonymization/index.md:
````diff
@@ -0,0 +1,221 @@
+python3 main.py config/yolo_train.yaml --yolo_train
+```
+
+#### Step 5: Anonymize Images in ROS2 Bag Files
````
```suggestion
#### Step 5: Anonymize Images in ROS 2 Bag Files
```

palas21 commented on docs/datasets/data-anonymization/index.md:
````diff
@@ -0,0 +1,221 @@
+You should set your configuration in config/yolo_create_dataset.yaml file.
+
+```bash
+python3 main.py config/yolo_create_dataset.yaml --yolo_create_dataset
````
YOLO config file is missing. Could you share a config file in release?

mitsudome-r replied:
It was set resolved, but I still got the following error when I tried to run this command:

```
Traceback (most recent call last):
  File "/home/mitsudome-r/autoware_rosbag2_anonymizer/main.py", line 81, in <module>
    yolo_create_dataset(
  File "/home/mitsudome-r/autoware_rosbag2_anonymizer/autoware_rosbag2_anonymizer/tools/yolo_create_dataset.py", line 28, in yolo_create_dataset
    unified_language_model = UnifiedLanguageModel(config_data, json_data)
  File "/home/mitsudome-r/autoware_rosbag2_anonymizer/autoware_rosbag2_anonymizer/model/unified_language_model.py", line 47, in __init__
    self.yolo_model_name = config["yolo"]["model"]
KeyError: 'yolo'
```

It seams like it is expecting yolo parameters defined in yolo_create_dataset.yaml, but it is not defined in the default configuration file.

StepTurtle replied:
I apologize, I missed that. I have added the missing variable and also included an `if` statement to check if the YOLO model exists. If users haven't downloaded the pre-trained model, it will still work.

related commit: https://github.com/autowarefoundation/autoware_rosbag2_anonymizer/commit/9ea65e1c8a9359626c95cba2f1df68102ebd645b

palas21 commented on docs/datasets/data-anonymization/index.md:
````diff
@@ -0,0 +1,221 @@
+
+```bash
+git clone https://github.com/autowarefoundation/autoware_rosbag2_anonymizer.git
+```
````
After clonning the repository user need to go to downloaded folder 
![image](https://github.com/autowarefoundation/autoware-documentation/assets/112626902/58223183-1d70-46e8-b697-2f29962f4e08)
For further downloads of pretrained models need to be in the autoware_rosbag2_anonymizer folder.

palas21 commented on docs/datasets/data-anonymization/index.md:
````diff
@@ -0,0 +1,221 @@
+You should set your configuration in config/yolo_anonymize.yaml file.
+
+```bash
+python3 main.py config/yolo_anonymize.yaml --yolo_anonymize
````
Some debug comments left on the 'autoware_rosbag2_anonymizer/tools/yolo_anonymize.py'.

StepTurtle replied:
```
0: 416x640 2 license plates, 151.4ms
Speed: 3.7ms preprocess, 151.4ms inference, 132.2ms postprocess per image at shape (1, 3, 640, 640)
```

If you mentioned about these outputs, these comes from YOLO and I thought these looks nice.

palas21 replied:
No, I mentioned about [these](https://github.com/autowarefoundation/autoware_rosbag2_anonymizer/blob/a87dd7a2cf59fb1883205a9203dc8607dd02e5a4/autoware_rosbag2_anonymizer/tools/yolo_anonymize.py#L65) debug comments. Also there are debug comments left on some other [[1]](https://github.com/autowarefoundation/autoware_rosbag2_anonymizer/blob/a87dd7a2cf59fb1883205a9203dc8607dd02e5a4/autoware_rosbag2_anonymizer/tools/anonymize_with_unified_model.py#L73) [[2]](https://github.com/autowarefoundation/autoware_rosbag2_anonymizer/blob/a87dd7a2cf59fb1883205a9203dc8607dd02e5a4/autoware_rosbag2_anonymizer/tools/yolo_create_dataset.py#L84C1-L97C29) files as well.

StepTurtle replied:
I have some future plan on these debug parts which are comment right now, I think it can be stay for now.

mitsudome-r commented:
I will test this tool this week.

mitsudome-r commented:
@StepTurtle 
I'm trying to follow the instructions, but I had to install the following additional dependencies  to run `python3 main.py config/anonymize_with_unified_model.yaml --anonymize_with_unified_model`:
- open_clip_torch
- groundingdino-py
- ultralytics

StepTurtle commented:
> @StepTurtle I'm trying to follow the instructions, but I had to install the following additional dependencies to run `python3 main.py config/anonymize_with_unified_model.yaml --anonymize_with_unified_model`:
> 
>     * open_clip_torch
> 
>     * groundingdino-py
> 
>     * ultralytics

Thanks for your reply, can you share output of the `python3 -m pip install .` command?

mitsudome-r commented:
@StepTurtle 
Here's the output:

```
Defaulting to user installation because normal site-packages is not writeable
Processing /home/mitsudome-r/autoware_rosbag2_anonymizer
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: cv-bridge in /opt/ros/humble/local/lib/python3.10/dist-packages (from autoware_rosbag2_anonymizer==0.1.0) (3.2.1)
Requirement already satisfied: huggingface-hub in /home/mitsudome-r/.local/lib/python3.10/site-packages (from autoware_rosbag2_anonymizer==0.1.0) (0.24.5)
Requirement already satisfied: numpy in /home/mitsudome-r/.local/lib/python3.10/site-packages (from autoware_rosbag2_anonymizer==0.1.0) (1.26.4)
Requirement already satisfied: opencv-python in /home/mitsudome-r/.local/lib/python3.10/site-packages (from autoware_rosbag2_anonymizer==0.1.0) (4.10.0.84)
Requirement already satisfied: Pillow in /home/mitsudome-r/.local/lib/python3.10/site-packages (from autoware_rosbag2_anonymizer==0.1.0) (10.4.0)
Requirement already satisfied: PyYAML in /home/mitsudome-r/.local/lib/python3.10/site-packages (from autoware_rosbag2_anonymizer==0.1.0) (6.0.1)
Requirement already satisfied: segment-anything in /home/mitsudome-r/.local/lib/python3.10/site-packages (from autoware_rosbag2_anonymizer==0.1.0) (1.0)
Requirement already satisfied: supervision>0.18.0 in /home/mitsudome-r/.local/lib/python3.10/site-packages (from autoware_rosbag2_anonymizer==0.1.0) (0.22.0)
Requirement already satisfied: torch in /home/mitsudome-r/.local/lib/python3.10/site-packages (from autoware_rosbag2_anonymizer==0.1.0) (2.4.0)
Requirement already satisfied: torchvision in /home/mitsudome-r/.local/lib/python3.10/site-packages (from autoware_rosbag2_anonymizer==0.1.0) (0.19.0)
Collecting open-clip-torch (from autoware_rosbag2_anonymizer==0.1.0)
  Using cached open_clip_torch-2.26.1-py3-none-any.whl.metadata (31 kB)
Collecting groundingdino-py (from autoware_rosbag2_anonymizer==0.1.0)
  Using cached groundingdino_py-0.4.0-py2.py3-none-any.whl
Requirement already satisfied: defusedxml<0.8.0,>=0.7.1 in /usr/lib/python3/dist-packages (from supervision>0.18.0->autoware_rosbag2_anonymizer==0.1.0) (0.7.1)
Requirement already satisfied: matplotlib>=3.6.0 in /home/mitsudome-r/.local/lib/python3.10/site-packages (from supervision>0.18.0->autoware_rosbag2_anonymizer==0.1.0) (3.9.0)
Requirement already satisfied: opencv-python-headless>=4.5.5.64 in /home/mitsudome-r/.local/lib/python3.10/site-packages (from supervision>0.18.0->autoware_rosbag2_anonymizer==0.1.0) (4.10.0.84)
Requirement already satisfied: scipy<2.0.0,>=1.10.0 in /home/mitsudome-r/.local/lib/python3.10/site-packages (from supervision>0.18.0->autoware_rosbag2_anonymizer==0.1.0) (1.12.0)
Requirement already satisfied: addict in /home/mitsudome-r/.local/lib/python3.10/site-packages (from groundingdino-py->autoware_rosbag2_anonymizer==0.1.0) (2.4.0)
Requirement already satisfied: pycocotools in /home/mitsudome-r/.local/lib/python3.10/site-packages (from groundingdino-py->autoware_rosbag2_anonymizer==0.1.0) (2.0.8)
INFO: pip is looking at multiple versions of groundingdino-py to determine which version is compatible with other requirements. This could take a while.
  Using cached groundingdino-py-0.3.0.tar.gz (82 kB)
  Preparing metadata (setup.py) ... done
  Using cached groundingdino-py-0.2.0.tar.gz (82 kB)
  Preparing metadata (setup.py) ... done
  Using cached groundingdino-py-0.1.2.tar.gz (82 kB)
  Preparing metadata (setup.py) ... done
  Using cached groundingdino-py-0.1.0.tar.gz (78 kB)
  Preparing metadata (setup.py) ... done
ERROR: Cannot install autoware-rosbag2-anonymizer==0.1.0 and autoware_rosbag2_anonymizer because these package versions have conflicting dependencies.

The conflict is caused by:
    autoware-rosbag2-anonymizer 0.1.0 depends on supervision>0.18.0
    groundingdino-py 0.4.0 depends on supervision==0.6.0
    autoware-rosbag2-anonymizer 0.1.0 depends on supervision>0.18.0
    groundingdino-py 0.3.0 depends on supervision==0.6.0
    autoware-rosbag2-anonymizer 0.1.0 depends on supervision>0.18.0
    groundingdino-py 0.2.0 depends on supervision==0.6.0
    autoware-rosbag2-anonymizer 0.1.0 depends on supervision>0.18.0
    groundingdino-py 0.1.2 depends on supervision==0.6.0
    autoware-rosbag2-anonymizer 0.1.0 depends on supervision>0.18.0
    groundingdino-py 0.1.0 depends on supervision==0.6.0

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip to attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
```

mitsudome-r commented:
Since the error was around supervision, I manually installed groundingdino, open_clip_torch, and ultralytics, then reinstalled supervision 0.22.0 manually.

StepTurtle commented:
> @StepTurtle I'm trying to follow the instructions, but I had to install the following additional dependencies to run `python3 main.py config/anonymize_with_unified_model.yaml --anonymize_with_unified_model`:
> 
>     * open_clip_torch
> 
>     * groundingdino-py
> 
>     * ultralytics

Fixed with this commit: https://github.com/autowarefoundation/autoware_rosbag2_anonymizer/commit/9445da4348904aa92603022bae1e3e7354b24b91

- I forked `groundingdino-py` library to leo-drive organization and I just changed the supervision version because upper version of `supervision`  is need from tool but any version is not necessary for `groundingdino-py`.

StepTurtle commented on docs/datasets/data-anonymization/index.md:
````diff
@@ -0,0 +1,231 @@
+### Download the pretrained models
+
+```bash
+wget https://dl.fbaipublicfiles.com/segment_anything/sam_vit_h_4b8939.pth
+
+wget https://huggingface.co/ShilongLiu/GroundingDINO/resolve/main/GroundingDINO_SwinB.cfg.py
+wget https://huggingface.co/ShilongLiu/GroundingDINO/resolve/main/groundingdino_swinb_cogcoor.pth
+
+wget https://github.com/autowarefoundation/autoware_rosbag2_anonymizer/releases/download/v0.0.0/yolov8x_anonymizer.pt
+wget https://github.com/autowarefoundation/autoware_rosbag2_anonymizer/releases/download/v0.0.0/yolo_config.yaml
````
I added citation information both [README.md](https://github.com/autowarefoundation/autoware_rosbag2_anonymizer/blob/main/README.md) and the end of the documentation.

StepTurtle commented on docs/datasets/data-anonymization/index.md:
```diff
@@ -0,0 +1,231 @@
+
+## Configuration
+
+Define prompts in the validation.jsonfile. The tool will use these prompts to detect objects. You can add your prompts
```
fixed.

mitsudome-r commented:
@StepTurtle Thanks for the ping. I'm doing another test to make sure that there are no issues left.

mitsudome-r commented on docs/datasets/data-anonymization/index.md:
````diff
@@ -0,0 +1,277 @@
+You should set your configuration in config/anonymize_with_unified_model.yaml file.
+
+```bash
+python3 main.py config/anonymize_with_unified_model.yaml --anonymize_with_unified_model
````
When I tried running this command, I got the following error message:

```
torch.OutOfMemoryError: CUDA out of memory. Tried to allocate 1024.00 MiB. GPU 0 has a total capacity of 10.87 GiB of which 1010.88 MiB is free. Including non-PyTorch memory, this process has 8.66 GiB memory in use. Of the allocated memory 8.21 GiB is allocated by PyTorch, and 266.44 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)
```

The error disappeared after setting the environment variable as suggested (`export PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True`). If this is expected, it might be user friendly to add in the instruction to set the variable first before running the command.

StepTurtle replied:
I knew this error was caused by the GPU running out of memory, but I didn't know it could be resolved this way. Thank you. I’ve added it under a new 'Troubleshooting' section.

StepTurtle replied:
btw, 'Troubleshooting' section add both package `readme.md` and documentation page.

mitsudome-r commented on docs/datasets/data-anonymization/index.md:
````diff
@@ -0,0 +1,292 @@
+python3 main.py config/yolo_create_dataset.yaml --yolo_create_dataset
+```
+
+#### Step 2: Manually Label the Missing Labels
+
+The dataset which is created in the first step has some missing labels. You should label the missing labels manually.
````
Is there a tool that we can use to add missing labels?

StepTurtle commented on docs/datasets/data-anonymization/index.md:
````diff
@@ -0,0 +1,292 @@
+You should set your configuration in config/anonymize_with_unified_model.yaml file.
+
+```bash
+python3 main.py config/anonymize_with_unified_model.yaml --anonymize_with_unified_model
````
Added a information to source workspace which includes custom message types.

related commits:
- https://github.com/autowarefoundation/autoware-documentation/pull/557/commits/d9a4c3e87643523045724731a20e6cf00e1ee072
- https://github.com/autowarefoundation/autoware_rosbag2_anonymizer/commit/0532ef245626b4fc794ee863bba552a6f18c23c3

mitsudome-r reviewed (approved):
LGTM!
